### ISSUE TITLE
Implement RANGE / ALIGN / FILL grammar

### ISSUE THREAD

**Issue by waynexia** [GreptimeTeam/greptimedb#1662]
### What problem does the new feature solve?

It's very useful if we can aggregate over time range like PromQL's range selector on time-series data.

### What does the feature do?

Add three new SQL keyword

- RANGE
- FILL
- ALIGN

### RANGE

Follow the function. It specifies the time range of data that feed into the range function. Range function should be a superset of window functions.

### FILL
Can be either global (at the end of SELECT statement) or local (follow the function). Specifies the default value when a range function doesn't have result (NULL)

###  ALIGN
Specifies the align step. Default align to calendar, or to the time the query is submitted.

## Examples

```
select tag1, rate(metrics) RANGE 5m, sum(metrics) FILL MAX, sum(metrics) RANGE 10m from t ALIGN 1h FILL NULL;
```
The above SQL has two range function `rate(metrics) RANGE 5m` and `sum(metrics) RANGE 10m`. One computes 5mins rate and one computes 10mins sum. A query can have different RANGE.

The result will be aligned to calendar with 1 hour step. I.e., the result will be at 00:00 a.m., 01:00 a.m etc.

**Comment by waynexia** [comment]
Related parser PR: https://github.com/GreptimeTeam/sqlparser-rs/pull/1

**Comment by Taylor-lagrange** [comment]
As of July 16, the things that have been completed are:
1. Update of sqlparser,  it can already recognize `fill` and `range` grammars
2. A simple `logicalplan rewriter` that can rewrite the logic plan generated by datafusion
3. A simple physical plan implementation, now you can input `SELECT avg(a) RANGE '5m' FROM test ALIGN '1h' FILL NULL;` from the command line and get a result. (Although the range-select algorithm is not implemented, the entire sql can run correctly with wrong output)

**Comment by Taylor-lagrange** [comment]
As of July 24, the things that have been completed are:

1. support calculation/nested expression rewriting | keyword `by` |  new `range_fn` rewrite logical in sqlparser
2. implement macro `DFConvert` to implement `From` Trait. Convert the current sqlparser struct to the struct used by datafusion sqlparser.

related pr https://github.com/GreptimeTeam/sqlparser-rs/pull/3

this week:

Implement a simple logical plan rewriter and a basical RangePlan

**Comment by waynexia** [comment]
Support to cast between two different versions of sqlparser https://github.com/GreptimeTeam/sqlparser-rs/pull/3

**Comment by Taylor-lagrange** [comment]
As of August 7, the things that have been completed are:

1. logical plan rewriter: fix code review, almost finish https://github.com/GreptimeTeam/greptimedb/pull/2058
2. fix format of `range_fn`, related pr: https://github.com/GreptimeTeam/sqlparser-rs/pull/4
3. Implementaion a physical plan, with unit test and sqlness test. Can execute arbitrary  nested/complex range queries from the cmd, range select can work normally https://github.com/Taylor-lagrange/greptimedb/commit/5549f6ab7ba047937334fe9f3c61ce12f007ed1d

this week:
1. merge logical plan rewriter/physical plan into codebase
2. optimize physical plan implementation

**Comment by tisonkun** [comment]
@waynexia @Taylor-lagrange  seems implemented?

**Comment by Taylor-lagrange** [comment]
Yes, the range query function has been implemented, and many bugs in range query have been fixed. Should this issue be closed? @waynexia

**Comment by waynexia** [comment]
Yes, this ticket is fully implemented. Closing it.
